plugins/nf-validation/src/main/nextflow/validation/SamplesheetConverter.groovy

```diff
@@ -32,251 +32,93 @@ import nextflow.Session
-            log.debug "subfield = $subfield"
-            def ArrayList result = input.collect{ castToNFType(it, subfield) } as ArrayList
             return result
         }
+        return channelFormat
+    }
 
-        def String inputStr = input as String
-        // Convert string values
-        if(type == "string" || !type) {
-            def String result = inputStr as String
-            
-            // Check and convert to the desired format
-            def String format = field['value']['format']
-            if(format) {
-                if(format == "file-path-pattern") {
-                    def ArrayList inputFiles = Nextflow.file(inputStr) as ArrayList
-                    return inputFiles
-                }
-                if(format.contains("path")) {
-                    def Path inputFile = Nextflow.file(inputStr) as Path
-                    return inputFile
+    // TODO add path casting
+    private static Object formatEntry(Object input, LinkedHashMap schema) {
+
+        if (input instanceof Map) {
+            def List result = []
+            def LinkedHashMap properties = schema["properties"]
+            properties.each { property, schemaValues ->
+                def value = input[property] ?: []
+                def List metaIds = schemaValues["meta"] instanceof List ? schemaValues["meta"] as List : schemaValues["meta"] instanceof String ? [schemaValues["meta"]] : []
+                if (metaIds) {
+                    metaIds.each {
+                        addMeta(["${it}":value])
+                    }
+                } else {
+                    result.add(formatEntry(value, schemaValues as LinkedHashMap))
                 }
             }
-            
-
-            // Return the plain string value
             return result
-        }
-
-        // Convert number values
-        else if(type == "number") {
-            try {
-                def int result = inputStr as int
-                return result
-            }
-            catch (NumberFormatException e) {
-                log.debug("Could not convert ${input} to an integer. Trying to convert to a float.")
-            }
-
-            try {
-                def float result = inputStr as float
-                return result
-            }
-            catch (NumberFormatException e) {
-                log.debug("Could not convert ${inputStr} to a float. Trying to convert to a double.")
+        } else if (input instanceof List) {
+            def List result = []
+            input.each {
+                result.add(formatEntry(it, schema["items"] as LinkedHashMap))
             }
-            
-            def double result = inputStr as double
             return result
+        } else {
+            def List formats = getPathFormats(schema)
+            if (formats && input instanceof String) {
+                return Nextflow.file(input)
+            }
+            return input
```

Comment:
I think this might break if `anyOf` or `oneOf`, using `file-path` (or other file-type formats) and some non-file-type e.g. `numeric` and when the supplied value is of the non-file type.

Reply:
It will only cast the item to a path if it's a string (which will not be the case if it is a `number`. I know this is not the best solution, but I don't really see a way without actually validating each possible subschema in the `anyOf` or `oneOf` sections. I'll take a better look when I have some time